draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1958 @@
+are the messages we publish to the :term:`Message Broker` whenever we want to
+execute a :term:`Task`.
+
+:term:`Document messages <Document Message>` are the messages we get as a result.
+
+:term:`Document messages <Document Message>` may also be produced whenever
+we publish a serialized representation of a :term:`Domain Model`.
+
+.. code-block:: pycon
+
+  >>> from celery import task
+  >>> @task
+  ... def add(a, b):
+  ...   return a + b
+  >>> result = add.delay(1, 2)  # Publish a command message
+  >>> result.get()  # Consume a Document message
+  3
+
+:term:`Event messages <Event Message>` are a new concept for Celery.
+They describe that a :term:`Domain Event` occurred.
+Multiple tasks can be subscribed to an event.
+
+.. code-block:: pycon
+
+  >>> from uuid import UUID
+  >>> from celery import task, event
+  >>> from myapp.models import User, AccountManager
+  >>> @task
+  ... def send_welcome_email(user_id, email):
+  ...   send_email(email=email, contents="hello, welcome", subject="welcome") # Send a welcome email to the user...
+  ...   User.objects.filter(pk=user_id).update(welcome_email_sent=True)
+  >>> @task
+  ... def notify_account_manager(user_id, email):
+  ...   account_manager = AccountManager.objects.assign_account_manager(user_id)
+  ...   send_email(email=account_manager.email, contents="you have a new user to attend to", subject="Alert") # Send an email to the account manager...
+  >>> @event
+  ... class UserRegistered:
+  ...   user_id: UUID
+  ...   email: str
+  >>> UserRegistered.subscribe(send_welcome_email)
+  >>> UserRegistered.subscribe(notify_account_manager)
+  >>> UserRegistered.delay(user_id=1, email='foo@bar.com')  # Calls both send_welcome_email and notify_account_manager with the provided arguments.
+
+These architectural building blocks will aid us in creating a better messaging
+system. To encourage `ubiquitous language`_, we will be using them in this
+document and in Celery 5's codebase as well.
+
+Message Passing Protocol
+------------------------
+
+In previous versions of Celery the only protocol we defined was for the message
+format itself.
+
+Whenever a client published a message to the broker using `task.delay()` Celery
+serialized it to a known format which it can use to invoke the task remotely.
+
+This meant that you needed a special client library to publish tasks that Celery
+will later consume and execute.
+The architectural implication is a coupling between the publisher and the consumer
+which defeats the general use-case of message passing: decoupling the publisher
+from the consumer.
+
+In addition, the communication protocol between the master process and the workers
+was an internal implementation detail which was not well defined or understood
+by the maintainers or users.
+
+Introduction to AMQP 1.0 Terminology
+++++++++++++++++++++++++++++++++++++
+
+AMQP 1.0 is a flexible messaging protocol that is designed to pass messages between nodes.
```

Comment:
this will lead rabbitmq users to use the amqp 1.0 plugin as rabbitmq do not support AMQP 1.0 by default right now. we should also keep both versions for BC.

Reply:
Celery will use AMQP 1.0 for internal communication.
We do not force the users to select a specific broker.

Reply:
great

Reply:
I think this feature can also be included in celery 4.x.x series

Reply:
I can't wait as that will let me use AWS AMQ as broker !! :)